[lldb] Add Model Context Protocol (MCP) support to LLDB

JDevlieghere opened the pull request:
This PR adds an MCP (Model Context Protocol ) server to LLDB. For motivation and background, please refer to the corresponding RFC: https://discourse.llvm.org/t/rfc-adding-mcp-support-to-lldb/86798

I implemented this as a new kind of plugin. The idea is that we could support multiple protocol servers  (e.g. if we want to support DAP from within LLDB). This also introduces a corresponding top-level command (`protocol-server`) with two subcommands to `start` and `stop` the server.

```
(lldb) protocol-server start MCP tcp://localhost:1234
MCP server started with connection listeners: connection://[::1]:1234, connection://[127.0.0.1]:1234
```

The MCP sever supports one tool (`lldb_command`) which executes a command, but can easily be extended with more commands.

JDevlieghere committed:
[lldb] Add MCP support to LLDB

https://discourse.llvm.org/t/rfc-adding-mcp-support-to-lldb/86798

JDevlieghere commented:
I'm adding the folks that chimed in on the RFC as reviewers.

ashgti commented on lldb/source/Plugins/Protocol/MCP/ProtocolServerMCP.cpp:
```diff
@@ -0,0 +1,280 @@
+
+  m_read_handles = std::move(*handles);
+  m_loop_thread = std::thread([=] {
+    llvm::set_thread_name("mcp-runloop");
```
For accounting, should this prefix the debugger name in case there are multiple debugger instances in the same process? Maybe `{debugger.name}.mcp.listener` (or `runloop`)?

ashgti commented on lldb/source/Plugins/Protocol/MCP/ProtocolServerMCP.cpp:
```diff
@@ -0,0 +1,280 @@
+  auto handles = m_listener->Accept(m_loop, [=](std::unique_ptr<Socket> sock) {
+    std::lock_guard<std::mutex> guard(m_server_mutex);
+
+    const std::string client_name =
+        llvm::formatv("client-{0}", m_clients.size() + 1).str();
```
For accounting, should this prefix the debugger name in case there are multiple debugger instances in the same process? Maybe `{debugger.name}.mcp.client-{i}`?

ashgti commented on lldb/source/Plugins/Protocol/MCP/ProtocolServerMCP.cpp:
```diff
@@ -0,0 +1,280 @@
+
+    lldb::IOObjectSP io(std::move(sock));
+
+    m_clients.emplace_back(io, [=]() {
+      llvm::set_thread_name(client_name + "-runloop");
+      if (auto Err = Run(std::make_unique<JSONRPCTransport>(io, io)))
+        LLDB_LOG_ERROR(GetLog(LLDBLog::Host), std::move(Err), "MCP Error: {0}");
+    });
```
I think it looks like we're only supporting connecting via sockets.

Instead of spawning a thread per-client, should we have the clients register their handles to the RunLoop and handle the requests that way? I think it would result in fewer threads. I think all the clients are sharing a connection to the same debugger instance and this may help prevent races between clients.

ashgti commented on lldb/source/Plugins/Protocol/MCP/Tool.cpp:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  CommandReturnObject result(/*colors=*/false);
+  m_debugger.GetCommandInterpreter().HandleCommand(arguments.c_str(),
```
Should we prevent some commands from being executed here?

For example, `quit`?

JDevlieghere replied:
I've added a FIXME for now.

ashgti commented on lldb/source/Plugins/Protocol/MCP/Protocol.h:
```diff
@@ -0,0 +1,131 @@
+llvm::json::Value toJSON(const Notification &);
+bool fromJSON(const llvm::json::Value &, Notification &, llvm::json::Path);
+
+struct ToolCapability {
+  bool listChanged = false;
+};
```
No changes needed, but reading up on the schema for this, I wonder if we could also support completion (https://modelcontextprotocol.io/specification/2025-03-26/server/utilities/completion#user-interaction-model). Maybe in the future that would be reasonable to implement.

JDevlieghere replied:
Yup, I think that's something we could support. Completion was added in the latest revision of the spec and to support older tools I've settled on the previous revision for now, but this might be a good motivator to move to the next version.

ashgti commented on lldb/source/Plugins/Protocol/MCP/Protocol.h:
```diff
@@ -0,0 +1,131 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
Could we include a top level comment with a link to 

https://github.com/modelcontextprotocol/modelcontextprotocol/blob/main/schema/2025-03-26/schema.json

For reference?

ashgti commented on lldb/source/Plugins/Protocol/MCP/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+add_lldb_library(lldbPluginProtocolServerMCP PLUGIN
```
Should this be behind a build flag?

I could see some distributions disabling this feature for security reasons.

Once an MCP server is running, it can run arbitrary commands in the debugger and may even be able to spawn arbitrary processes.

ashgti reviewed (commented):
Would it be worth it to include a file in lldb/docs/resources about running this? For folks wanting to get started trying this out?

Maybe in a follow up PR.


Also, thinking about the overall flow of the MCP server and the debugger instance.

I am curious how this would work with lldb-dap. In the DAP, we make a debugger instance for each debug session and it goes away at the end of the debug session. So, I may not be able to run both lldb-dap and an MCP server in the same debug session, or at least the DAP session would stop the MCP server once the debugger session is over.

Should the MCP server mode have some way of creating a debugger instance itself? Like, should each client have its own debugger instance? If so, that sort of inverts how this should be arranged such that the MCP server should own the debugger not the other way around.

labath commented on lldb/unittests/Protocol/ProtocolMCPServerTest.cpp:
```diff
@@ -0,0 +1,195 @@
+  std::unique_ptr<TestJSONTransport> m_transport_up;
+  std::unique_ptr<TestProtocolServerMCP> m_server_up;
+
+  static constexpr llvm::StringLiteral k_ip = "127.0.0.1";
```
can you make this "localhost"? We actually have servers with no IPv4 stack o_O

labath reviewed (commented):
A very lightweight review. I'm focusing on the low-level details. I have no clue about this AI stuff.

JDevlieghere commented:
- Added a CMake flag to disable building protocol servers at configuration time. 
 - Added a link to the spec, settled on 2024-11-05 as that's what Claude Desktop supports right now and we weren't using anything from 2025-03-26.
 - Eliminated all threads except the one that runs the MCP MainLoop, which now serves all the clients.
 - Correctly handle notifications instead of treating them like requests.

ashgti commented on lldb/source/Plugins/Protocol/MCP/Protocol.cpp:
```diff
@@ -0,0 +1,203 @@
+    return true;
+  }
+
+  // Treat every other message as a request.
+  protocol::Request R;
+  if (!fromJSON(V, R, P))
+    return false;
+  M = std::move(R);
+  return true;
```
I think there is one other case for `Response` types.

If the `result` field exists, then its a `Response`.

Also, the `Request` requires the `method` field to exist, otherwise we have a malformed request.

ashgti commented on lldb/source/Plugins/Protocol/MCP/Protocol.cpp:
```diff
@@ -0,0 +1,203 @@
+    return false;
+  }
+
+  if (!O->get("jsonrpc")) {
+    P.report("not a valid JSON RPC message");
+    return false;
+  }
```
We should also check that the value is `"2.0"` otherwise it may be a different version of the json-rpc spec.

ashgti commented on lldb/unittests/Protocol/ProtocolMCPServerTest.cpp:
```diff
@@ -0,0 +1,195 @@
+  }
+};
+
+class ProtocolServerMCPTest : public ::testing::Test {
```
I think we also need a test for the `notifications/initialized` notification.

ashgti commented on lldb/source/Plugins/Protocol/MCP/Tool.h:
```diff
@@ -0,0 +1,55 @@
+  Tool(std::string name, std::string description);
+  virtual ~Tool() = default;
+
+  virtual protocol::TextResult Call(const llvm::json::Value &args) = 0;
```
Should the tools support returning an error? Should this be an `llvm::Expected<protocol::TextResult>`?

ashgti commented on lldb/source/Plugins/Protocol/MCP/Tool.cpp:
```diff
@@ -0,0 +1,70 @@
+
+protocol::TextResult LLDBCommandTool::Call(const llvm::json::Value &args) {
+  std::string arguments;
+  if (const json::Object *args_obj = args.getAsObject()) {
+    if (const json::Value *s = args_obj->get("arguments")) {
+      arguments = s->getAsString().value_or("");
+    }
+  }
```
Should we add a helper for decoding the arguments the LLDB command expects? We have the schema specified, so I think we could define this type at least internally to this file.

JDevlieghere commented:
Rebased and addressed @ashgti's comments:

- Support decoding `Response` messages
- Parse the JSON RPC version
- Test notification handlers
- Support returning errors from tool calls
- Decode LLDBTool arguments into a POD type

ashgti reviewed (approved):
LGTM!

I look forward to trying this out and iterating on this.

ashgti commented on lldb/source/Commands/CommandObjectProtocolServer.cpp:
```diff
@@ -0,0 +1,186 @@
+validateConnection(llvm::StringRef conn) {
+  auto uri = lldb_private::URI::Parse(conn);
+
+  if (uri && (uri->scheme == "tcp" || uri->scheme == "connect" ||
```
I'd rather them be consistent if we need to change lldb-dap as well.

JDevlieghere commented on lldb/source/Plugins/Protocol/MCP/ProtocolServerMCP.cpp:
```diff
@@ -0,0 +1,280 @@
+
+  // Wait for all our clients to exit.
+  for (auto &client : m_clients) {
+    client.first->Close();
```
https://github.com/llvm/llvm-project/pull/144610
